<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Basic Read Alignment</title>
  <meta name="description" content="A minimal guide to performing read alignment against a reference" />
  <meta name="generator" content="bookdown 0.13 and GitBook 2.6.7" />

  <meta property="og:title" content="Basic Read Alignment" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A minimal guide to performing read alignment against a reference" />
  <meta name="github-repo" content="TeamMacLean/basic_alignment" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Basic Read Alignment" />
  
  <meta name="twitter:description" content="A minimal guide to performing read alignment against a reference" />
  

<meta name="author" content="Dan MacLean" />


<meta name="date" content="2020-01-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="intro.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />










<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A minimal alignment guide</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About this course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a><ul>
<li class="chapter" data-level="1.1.1" data-path="index.html"><a href="index.html#local-software-installation"><i class="fa fa-check"></i><b>1.1.1</b> Local Software Installation</a></li>
<li class="chapter" data-level="1.1.2" data-path="index.html"><a href="index.html#sample-reference-genome-and-reads"><i class="fa fa-check"></i><b>1.1.2</b> Sample reference genome and reads</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> The Alignment Pipeline</a><ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#align-all-reads-to-a-reference"><i class="fa fa-check"></i><b>2.1</b> Align all reads to a reference</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#filter-out-badly-scoring-reads"><i class="fa fa-check"></i><b>2.2</b> Filter out badly scoring reads</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#sort-and-compress-output"><i class="fa fa-check"></i><b>2.3</b> Sort and compress output</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#further-reading"><i class="fa fa-check"></i><b>2.4</b> Further Reading</a><ul>
<li class="chapter" data-level="2.4.1" data-path="intro.html"><a href="intro.html#fastq-quality-scores"><i class="fa fa-check"></i><b>2.4.1</b> FastQ quality scores</a></li>
<li class="chapter" data-level="2.4.2" data-path="intro.html"><a href="intro.html#why-minimap2-and-not-bwabowtieother"><i class="fa fa-check"></i><b>2.4.2</b> Why minimap2 and not bwa|bowtie|other</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="running-minimap2.html"><a href="running-minimap2.html"><i class="fa fa-check"></i><b>3</b> Running <code>minimap2</code></a><ul>
<li class="chapter" data-level="3.1" data-path="running-minimap2.html"><a href="running-minimap2.html#the-minimap2-command-and-options"><i class="fa fa-check"></i><b>3.1</b> The <code>minimap2</code> command and options</a></li>
<li class="chapter" data-level="3.2" data-path="running-minimap2.html"><a href="running-minimap2.html#further-reading-1"><i class="fa fa-check"></i><b>3.2</b> Further Reading</a><ul>
<li class="chapter" data-level="3.2.1" data-path="running-minimap2.html"><a href="running-minimap2.html#the-operator"><i class="fa fa-check"></i><b>3.2.1</b> The <code>&gt;</code> operator</a></li>
<li class="chapter" data-level="3.2.2" data-path="running-minimap2.html"><a href="running-minimap2.html#minimap2-further-instructions-and-github"><i class="fa fa-check"></i><b>3.2.2</b> <code>minimap2</code> further instructions and github</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html"><i class="fa fa-check"></i><b>4</b> Filtering Badly Aligned Reads</a><ul>
<li class="chapter" data-level="4.1" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#sam-format"><i class="fa fa-check"></i><b>4.1</b> SAM Format</a></li>
<li class="chapter" data-level="4.2" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#samtools"><i class="fa fa-check"></i><b>4.2</b> <code>samtools</code></a></li>
<li class="chapter" data-level="4.3" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#the-samtools-command-and-options"><i class="fa fa-check"></i><b>4.3</b> The <code>samtools</code> command and options</a></li>
<li class="chapter" data-level="4.4" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#checking-the-filtering"><i class="fa fa-check"></i><b>4.4</b> Checking the filtering</a></li>
<li class="chapter" data-level="4.5" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#are-we-done"><i class="fa fa-check"></i><b>4.5</b> Are we done?</a></li>
<li class="chapter" data-level="4.6" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#further-reading-2"><i class="fa fa-check"></i><b>4.6</b> Further Reading</a><ul>
<li class="chapter" data-level="4.6.1" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#sam-format-1"><i class="fa fa-check"></i><b>4.6.1</b> SAM Format</a></li>
<li class="chapter" data-level="4.6.2" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#mapping-quality"><i class="fa fa-check"></i><b>4.6.2</b> Mapping Quality</a></li>
<li class="chapter" data-level="4.6.3" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#flags"><i class="fa fa-check"></i><b>4.6.3</b> Flags</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html"><i class="fa fa-check"></i><b>5</b> Connecting Programs and Compressing output</a><ul>
<li class="chapter" data-level="5.1" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#bam-files"><i class="fa fa-check"></i><b>5.1</b> BAM Files</a></li>
<li class="chapter" data-level="5.2" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#connecting-program-input-and-output-with-pipes"><i class="fa fa-check"></i><b>5.2</b> Connecting Program Input and Output With Pipes</a></li>
<li class="chapter" data-level="5.3" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#from-reads-to-filtered-alignments-in-one-step"><i class="fa fa-check"></i><b>5.3</b> From reads to filtered alignments in one step</a><ul>
<li class="chapter" data-level="5.3.1" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#from-reads-to-filtered-alignments-in-a-bam-file-in-one-step"><i class="fa fa-check"></i><b>5.3.1</b> From reads to filtered alignments in a BAM file in one step</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#sorting-bam-files"><i class="fa fa-check"></i><b>5.4</b> Sorting BAM files</a><ul>
<li class="chapter" data-level="5.4.1" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#automatically-deleting-the-unsorted-bam"><i class="fa fa-check"></i><b>5.4.1</b> Automatically deleting the unsorted BAM</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#indexing-the-sorted-bam"><i class="fa fa-check"></i><b>5.5</b> Indexing the sorted BAM</a></li>
<li class="chapter" data-level="5.6" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#further-reading-3"><i class="fa fa-check"></i><b>5.6</b> Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="automating-the-process.html"><a href="automating-the-process.html"><i class="fa fa-check"></i><b>6</b> Automating The Process</a><ul>
<li class="chapter" data-level="6.1" data-path="automating-the-process.html"><a href="automating-the-process.html#shell-scripts"><i class="fa fa-check"></i><b>6.1</b> Shell scripts</a><ul>
<li class="chapter" data-level="6.1.1" data-path="automating-the-process.html"><a href="automating-the-process.html#using-nano-to-create-a-shell-script"><i class="fa fa-check"></i><b>6.1.1</b> Using <code>nano</code> to create a shell script</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="automating-the-process.html"><a href="automating-the-process.html#creating-a-script-that-automates-our-alignment-pipeline."><i class="fa fa-check"></i><b>6.2</b> Creating a script that automates our alignment pipeline.</a></li>
<li class="chapter" data-level="6.3" data-path="automating-the-process.html"><a href="automating-the-process.html#running-the-script"><i class="fa fa-check"></i><b>6.3</b> Running the script</a></li>
<li class="chapter" data-level="6.4" data-path="automating-the-process.html"><a href="automating-the-process.html#running-on-different-input-files"><i class="fa fa-check"></i><b>6.4</b> Running on different input files</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html"><i class="fa fa-check"></i><b>7</b> Running an alignment on the HPC</a><ul>
<li class="chapter" data-level="7.1" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#an-hpc-is-a-group-of-slave-computers-under-control-of-a-master-computer"><i class="fa fa-check"></i><b>7.1</b> An HPC is a group of slave computers under control of a master computer</a></li>
<li class="chapter" data-level="7.2" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#logging-into-the-submission-node"><i class="fa fa-check"></i><b>7.2</b> Logging into the submission node</a></li>
<li class="chapter" data-level="7.3" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#preparing-a-job"><i class="fa fa-check"></i><b>7.3</b> Preparing a job</a><ul>
<li class="chapter" data-level="7.3.1" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#the-sbatch-options"><i class="fa fa-check"></i><b>7.3.1</b> The <code>#SBATCH</code> options</a></li>
<li class="chapter" data-level="7.3.2" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#the-source-options"><i class="fa fa-check"></i><b>7.3.2</b> The <code>source</code> options</a></li>
<li class="chapter" data-level="7.3.3" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#the-srun-command"><i class="fa fa-check"></i><b>7.3.3</b> The <code>srun</code> command</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#submitting-with-sbatch"><i class="fa fa-check"></i><b>7.4</b> Submitting with <code>sbatch</code></a></li>
<li class="chapter" data-level="7.5" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#checkout-tasks"><i class="fa fa-check"></i><b>7.5</b> Checkout tasks</a></li>
<li class="chapter" data-level="7.6" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#further-reading-4"><i class="fa fa-check"></i><b>7.6</b> Further Reading</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Basic Read Alignment</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Basic Read Alignment</h1>
<p class="author"><em>Dan MacLean</em></p>
<p class="date"><em>2020-01-14</em></p>
</div>
<div id="about-this-course" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> About this course</h1>
<p>In this short course we’ll look at a method for running an alignment of sequence reads against a reference genome. The course is very brief and will show you how to use a program called <code>minimap2</code> and <code>samtools</code> to create a binary alignment file that you can use in further work.</p>
<p>I acknowledge that there are lots of other programs and methods - this course is <em>not</em> meant to be comprehensive, it is meant to get you being productive. Seek out further advice if you need to run other programs. Do be encouraged though, 99 % of what you learn here will be applicable to other tools for the same job (they all run in a <em>very</em> similar manner) so this is a good place to start.</p>
<p>The course has two main parts - first we’ll learn to do this ‘locally’, that is to say on the computer that you are actually physically sitting at and have direct control over. Once we’ve done that and know how to run the actual programs then we shall switch to running the programs ‘remotely’ on a HPC environment using a submission system that you have to log in to.</p>
<div id="prerequisites" class="section level2">
<h2><span class="header-section-number">1.1</span> Prerequisites</h2>
<p>This course assumes that you are familiar with the basics of running stuff from a command-line. You’ll need to have some experience not lots. If you’ve done the TSL command-line course you’ll know plenty</p>
<p>For the first ‘local’ part of this you’ll need some software on your machines. Most bioinformatics software has to run on Unix style computers, which for most of us means Macs. The installation instructions below only apply to Macs.</p>
<p>For the second part you’ll need an HPC account. See the Bioinformatics Team to get one of these.</p>
<div id="local-software-installation" class="section level3">
<h3><span class="header-section-number">1.1.1</span> Local Software Installation</h3>
<p>We need to install <code>minimap2</code> and <code>samtools</code>. Installing bioinformatics software is often <em>not</em> straightforward so we’ll take a path of least resistance and install some tools that manage software for us. This is a roundabout way of doing things, but it greatly simplifies the hard parts and means that we can isolate our installations from the rest of our computer and not mess up anything already installed.</p>
<ol style="list-style-type: decimal">
<li><strong>Get <code>conda</code></strong>
<code>conda</code> is an environment and package manager for software projects (initially Python - hence the name). Its purpose is to create a sandbox area on your computer where you can safely install software without it interfering or overwriting any of the existing software already on there. This safe space is called a conda ‘environment’. To install <code>conda</code>:
<ol style="list-style-type: decimal">
<li>Click this link <a href="https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.pkg">https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.pkg</a> and wait for the package to download. When it has double-click it and go through the installation process.</li>
</ol></li>
<li><strong>Get <code>bioconda</code></strong>
<code>bioconda</code> is plugin that makes <code>conda</code> aware of the bioinformatics software we will need. To make <code>conda</code> aware of <code>bioconda</code>
<ol style="list-style-type: decimal">
<li>Open <code>Terminal</code> and type the following:</li>
</ol></li>
</ol>
<pre><code>      conda config --add channels defaults
      conda config --add channels bioconda
      conda config --add channels conda-forge</code></pre>
<ol start="3" style="list-style-type: decimal">
<li><strong>Create a new environment</strong>
We can now create our new safe space environment. Type in the following.</li>
</ol>
<pre><code>      conda create --name alignment_env</code></pre>
<p>This step can take a moment or two and nothing seems to happen for a while - be patient. Accept the defaults (answer <code>y</code> when questioned).</p>
<ol start="4" style="list-style-type: decimal">
<li><strong>Activate the environment</strong>
Everytime we want to use the new environment, including to install something into it, we have to activate it. That means each time we leave or start a new Terminal we need to type this into it</li>
</ol>
<pre><code>  conda activate alignment_env</code></pre>
<p>You should see the name at the <code>$</code> prompt change, telling you that this Terminal is in the <code>alignment_env</code> environment. If you ever can’t find programs that you’re sure you installed, it means that you probably didn’t activate the right environment.</p>
<ol start="5" style="list-style-type: decimal">
<li><strong>Install the alignment software <code>minimap2</code> and <code>samtools</code></strong>
All the steps up to now have been so we can do this one! Install each of the pieces of software in turn by using <code>conda</code> with:</li>
</ol>
<pre><code>      conda install minimap2
      conda install samtools</code></pre>
<p>These steps can also take a while. Again accept the defaults (answer <code>y</code>).</p>
<p>Now you are done! Everything is installed ready for you to work with. Next we need to get the sample data</p>
</div>
<div id="sample-reference-genome-and-reads" class="section level3">
<h3><span class="header-section-number">1.1.2</span> Sample reference genome and reads</h3>
<p>You’ll need this zip file of data: <a href="https://github.com/TeamMacLean/basic_alignment/blob/master/sample_data/sample_data.zip">sample_data.zip</a> which contains a reference genome and some file and a set of paired end reads <a href="https://github.com/TeamMacLean/basic_alignment/blob/master/sample_data/ecoli_left_R1.fq">left read</a> <a href="https://github.com/TeamMacLean/basic_alignment/blob/master/sample_data/ecoli_right_R2.fq">right read</a> . Download it, extract the files and put them into a folder on your machine. I suggest something like <code>Desktop/align_tut</code>. This will be the directory we’ll work from in the rest of the course.</p>
<p>That’s all you need to do the lesson. If you have any problems getting this going, then ask someone in the Bioinformatics Team and we’ll help.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="intro.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["Basic Read Alignment.pdf", "Basic Read Alignment.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
