<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 The Alignment Pipeline | Basic Read Alignment</title>
  <meta name="description" content="A minimal guide to performing read alignment against a reference" />
  <meta name="generator" content="bookdown 0.13 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 The Alignment Pipeline | Basic Read Alignment" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A minimal guide to performing read alignment against a reference" />
  <meta name="github-repo" content="TeamMacLean/basic_alignment" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 The Alignment Pipeline | Basic Read Alignment" />
  
  <meta name="twitter:description" content="A minimal guide to performing read alignment against a reference" />
  

<meta name="author" content="Dan MacLean" />


<meta name="date" content="2019-12-03" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="running-minimap2.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />










<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A minimal alignment guide</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About this course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a><ul>
<li class="chapter" data-level="1.1.1" data-path="index.html"><a href="index.html#local-software-installation"><i class="fa fa-check"></i><b>1.1.1</b> Local Software Installation</a></li>
<li class="chapter" data-level="1.1.2" data-path="index.html"><a href="index.html#sample-reference-genome-and-reads"><i class="fa fa-check"></i><b>1.1.2</b> Sample reference genome and reads</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> The Alignment Pipeline</a><ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#align-all-reads-to-a-reference"><i class="fa fa-check"></i><b>2.1</b> Align all reads to a reference</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#filter-out-badly-scoring-reads"><i class="fa fa-check"></i><b>2.2</b> Filter out badly scoring reads</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#sort-and-compress-output"><i class="fa fa-check"></i><b>2.3</b> Sort and compress output</a></li>
<li class="chapter" data-level="2.4" data-path="intro.html"><a href="intro.html#further-reading"><i class="fa fa-check"></i><b>2.4</b> Further Reading</a><ul>
<li class="chapter" data-level="2.4.1" data-path="intro.html"><a href="intro.html#fastq-quality-scores"><i class="fa fa-check"></i><b>2.4.1</b> FastQ quality scores</a></li>
<li class="chapter" data-level="2.4.2" data-path="intro.html"><a href="intro.html#why-minimap2-and-not-bwabowtieother"><i class="fa fa-check"></i><b>2.4.2</b> Why minimap2 and not bwa|bowtie|other</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="running-minimap2.html"><a href="running-minimap2.html"><i class="fa fa-check"></i><b>3</b> Running <code>minimap2</code></a><ul>
<li class="chapter" data-level="3.1" data-path="running-minimap2.html"><a href="running-minimap2.html#the-minimap2-command-and-options"><i class="fa fa-check"></i><b>3.1</b> The <code>minimap2</code> command and options</a></li>
<li class="chapter" data-level="3.2" data-path="running-minimap2.html"><a href="running-minimap2.html#further-reading-1"><i class="fa fa-check"></i><b>3.2</b> Further Reading</a><ul>
<li class="chapter" data-level="3.2.1" data-path="running-minimap2.html"><a href="running-minimap2.html#the-operator"><i class="fa fa-check"></i><b>3.2.1</b> The <code>&gt;</code> operator</a></li>
<li class="chapter" data-level="3.2.2" data-path="running-minimap2.html"><a href="running-minimap2.html#minimap2-further-instructions-and-github"><i class="fa fa-check"></i><b>3.2.2</b> <code>minimap2</code> further instructions and github</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html"><i class="fa fa-check"></i><b>4</b> Filtering Badly Aligned Reads</a><ul>
<li class="chapter" data-level="4.1" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#sam-format"><i class="fa fa-check"></i><b>4.1</b> SAM Format</a></li>
<li class="chapter" data-level="4.2" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#samtools"><i class="fa fa-check"></i><b>4.2</b> <code>samtools</code></a></li>
<li class="chapter" data-level="4.3" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#the-samtools-command-and-options"><i class="fa fa-check"></i><b>4.3</b> The <code>samtools</code> command and options</a></li>
<li class="chapter" data-level="4.4" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#checking-the-filtering"><i class="fa fa-check"></i><b>4.4</b> Checking the filtering</a></li>
<li class="chapter" data-level="4.5" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#are-we-done"><i class="fa fa-check"></i><b>4.5</b> Are we done?</a></li>
<li class="chapter" data-level="4.6" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#further-reading-2"><i class="fa fa-check"></i><b>4.6</b> Further Reading</a><ul>
<li class="chapter" data-level="4.6.1" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#sam-format-1"><i class="fa fa-check"></i><b>4.6.1</b> SAM Format</a></li>
<li class="chapter" data-level="4.6.2" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#mapping-quality"><i class="fa fa-check"></i><b>4.6.2</b> Mapping Quality</a></li>
<li class="chapter" data-level="4.6.3" data-path="filtering-badly-aligned-reads.html"><a href="filtering-badly-aligned-reads.html#flags"><i class="fa fa-check"></i><b>4.6.3</b> Flags</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html"><i class="fa fa-check"></i><b>5</b> Connecting Programs and Compressing output</a><ul>
<li class="chapter" data-level="5.1" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#bam-files"><i class="fa fa-check"></i><b>5.1</b> BAM Files</a></li>
<li class="chapter" data-level="5.2" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#connecting-program-input-and-output-with-pipes"><i class="fa fa-check"></i><b>5.2</b> Connecting Program Input and Output With Pipes</a></li>
<li class="chapter" data-level="5.3" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#from-reads-to-filtered-alignments-in-one-step"><i class="fa fa-check"></i><b>5.3</b> From reads to filtered alignments in one step</a><ul>
<li class="chapter" data-level="5.3.1" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#from-reads-to-filtered-alignments-in-a-bam-file-in-one-step"><i class="fa fa-check"></i><b>5.3.1</b> From reads to filtered alignments in a BAM file in one step</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#sorting-bam-files"><i class="fa fa-check"></i><b>5.4</b> Sorting BAM files</a><ul>
<li class="chapter" data-level="5.4.1" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#automatically-deleting-the-unsorted-bam"><i class="fa fa-check"></i><b>5.4.1</b> Automatically deleting the unsorted BAM</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#indexing-the-sorted-bam"><i class="fa fa-check"></i><b>5.5</b> Indexing the sorted BAM</a></li>
<li class="chapter" data-level="5.6" data-path="connecting-programs-and-compressing-output.html"><a href="connecting-programs-and-compressing-output.html#further-reading-3"><i class="fa fa-check"></i><b>5.6</b> Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="automating-the-process.html"><a href="automating-the-process.html"><i class="fa fa-check"></i><b>6</b> Automating The Process</a><ul>
<li class="chapter" data-level="6.1" data-path="automating-the-process.html"><a href="automating-the-process.html#shell-scripts"><i class="fa fa-check"></i><b>6.1</b> Shell scripts</a><ul>
<li class="chapter" data-level="6.1.1" data-path="automating-the-process.html"><a href="automating-the-process.html#using-nano-to-create-a-shell-script"><i class="fa fa-check"></i><b>6.1.1</b> Using <code>nano</code> to create a shell script</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="automating-the-process.html"><a href="automating-the-process.html#creating-a-script-that-automates-our-alignment-pipeline."><i class="fa fa-check"></i><b>6.2</b> Creating a script that automates our alignment pipeline.</a></li>
<li class="chapter" data-level="6.3" data-path="automating-the-process.html"><a href="automating-the-process.html#running-the-script"><i class="fa fa-check"></i><b>6.3</b> Running the script</a></li>
<li class="chapter" data-level="6.4" data-path="automating-the-process.html"><a href="automating-the-process.html#running-on-different-input-files"><i class="fa fa-check"></i><b>6.4</b> Running on different input files</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html"><i class="fa fa-check"></i><b>7</b> Running an alignment on the HPC</a><ul>
<li class="chapter" data-level="7.1" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#an-hpc-is-a-group-of-slave-computers-under-control-of-a-master-computer"><i class="fa fa-check"></i><b>7.1</b> An HPC is a group of slave computers under control of a master computer</a></li>
<li class="chapter" data-level="7.2" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#logging-into-the-submission-node"><i class="fa fa-check"></i><b>7.2</b> Logging into the submission node</a></li>
<li class="chapter" data-level="7.3" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#preparing-a-job"><i class="fa fa-check"></i><b>7.3</b> Preparing a job</a><ul>
<li class="chapter" data-level="7.3.1" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#the-sbatch-options"><i class="fa fa-check"></i><b>7.3.1</b> The <code>#SBATCH</code> options</a></li>
<li class="chapter" data-level="7.3.2" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#the-source-options"><i class="fa fa-check"></i><b>7.3.2</b> The <code>source</code> options</a></li>
<li class="chapter" data-level="7.3.3" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#the-srun-command"><i class="fa fa-check"></i><b>7.3.3</b> The <code>srun</code> command</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#submitting-with-sbatch"><i class="fa fa-check"></i><b>7.4</b> Submitting with <code>sbatch</code></a></li>
<li class="chapter" data-level="7.5" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#checkout-tasks"><i class="fa fa-check"></i><b>7.5</b> Checkout tasks</a></li>
<li class="chapter" data-level="7.6" data-path="running-an-alignment-on-the-hpc.html"><a href="running-an-alignment-on-the-hpc.html#further-reading-4"><i class="fa fa-check"></i><b>7.6</b> Further Reading</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Basic Read Alignment</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="intro" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> The Alignment Pipeline</h1>
<p>In this chapter we’ll look at an overview standard paired-end read run of <code>minimap2</code>, what it outputs and how to manipulate the output with <code>samtools</code>.</p>
<p>The overall pipeline is very straightforward…</p>
<ol style="list-style-type: decimal">
<li>Align all reads to a reference</li>
<li>Filter badly scoring reads</li>
<li>Sort and compress output</li>
</ol>
<div id="align-all-reads-to-a-reference" class="section level2">
<h2><span class="header-section-number">2.1</span> Align all reads to a reference</h2>
<p>This is the main step, and with <code>minimap2</code> it can be accomplished with a single command-line. In this step each read is aligned against the reference, and its best aligning position found. That position, along with a metric of the quality of the single alignment is reported in a <code>SAM</code> format file.</p>
</div>
<div id="filter-out-badly-scoring-reads" class="section level2">
<h2><span class="header-section-number">2.2</span> Filter out badly scoring reads</h2>
<p>This is the quality control step. We remove reads that don’t have a good alignment score because in most contexts it means the read is a bad read with bad sequence in it. Of course in some contexts it isn’t - it depends what you’re aligning to what, but for the RNAseq situation or SNP calling situation where we expect the reads to be very like the reference this is appropriate. This step is done with <code>samtools</code></p>
</div>
<div id="sort-and-compress-output" class="section level2">
<h2><span class="header-section-number">2.3</span> Sort and compress output</h2>
<p>Once filtering is done and we have the set of reads we wish to retain we can take our output file and convert it to a sorted binary format that uses less disk space and is optimised for searching in downstream analysis. This step is a kind of housekeeping step that makes everything later easier. We do it with <code>samtools</code></p>
</div>
<div id="further-reading" class="section level2">
<h2><span class="header-section-number">2.4</span> Further Reading</h2>
<div id="fastq-quality-scores" class="section level3">
<h3><span class="header-section-number">2.4.1</span> FastQ quality scores</h3>
<p>Typically the reads we use will be in the FastQ format, this is a two line format that encodes not only the sequence read, but the ‘believability’ of each base. You can see more <a href="https://en.wikipedia.org/wiki/FASTQ_format">here</a></p>
</div>
<div id="why-minimap2-and-not-bwabowtieother" class="section level3">
<h3><span class="header-section-number">2.4.2</span> Why minimap2 and not bwa|bowtie|other</h3>
<p>You may have heard of other aligners and that people in your group are using them. <code>minimap2</code> is currently the quickest and most accurate of all the aligners and has general ability with all modern sequence types. Those using others (in particular bwa and bowtie) are pretty out of date. bwa <em>was</em> a favourite for a long time, but has been superceded. The same author created both bwa and minimap2 and minimap2 was written to address issues with bwa.</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="running-minimap2.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["Basic Read Alignment.pdf", "Basic Read Alignment.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
